### ISSUE TITLE
Found also non CupsPrinter

### ISSUE THREAD

**Issue by chrisapps4u** [BenoitDuffez/AndroidCupsPrint#118]
![find nocups printer](https://user-images.githubusercontent.com/43114622/45768242-a493a800-bc3c-11e8-9543-824784520907.JPG)

On this picture you can see, that the MdnsServices also find printers that are not printers of a CupsServer (the Photosmart Prem C410). The reason is, that we searching for all ipp printers. So all network printer answers. But we get an error if we try to print to them. So we should check, if the printers we get, are printers of a CupsServer. The only thing come in my mind is, to ask again the CupsServer for his printers. I will make an example for that.

**Comment by chrisapps4u** [comment]
see: [https://github.com/chrisapps4u/AndroidCupsPrint/commit/6ab99acc27d4d0640e92c855a96287239541b937](https://github.com/chrisapps4u/AndroidCupsPrint/commit/6ab99acc27d4d0640e92c855a96287239541b937)

Then we also save the server of the ipp printers and ask again the server for cups pinters. All network printer who don't are a cups server are excluded now.

**Comment by BenoitDuffez** [comment]
The idea is nice, but what confuses me is that the app should work with IPP printers too. Isn't that the case?

**Comment by chrisapps4u** [comment]
No, it dosn't work with IPP printers. But the printers are shown by the MsdnServices.
The following document shows a list of the my mDNS-Services with avahi-browse: [avahi-domains.txt](https://github.com/BenoitDuffez/AndroidCupsPrint/files/2408991/avahi-domains.txt)
There you can see that the "Photosmart Prem C410 series" also make a response on port 631. That is detected by the MsdnServices and finally shown as a printer of AndroidCupsPrint. But this is wrong. We shouldn't show these printers, because the don't work with AndroidCupsPrint.

**Comment by chrisapps4u** [comment]
In this case i have another idea. We could look in MsdnServices if the printer have the arguments `printer-type` and `printer-state`. This arguments seems to be only send by a cups printer. But i don't find any information of the [CUPS Implementation](https://www.cups.org/doc/spec-ipp.html), which mDNS information should be send.

**Comment by BenoitDuffez** [comment]
What do you mean it doesn't work with IPP printers? On my setup, the mDNS returns the only printer of the network, and the URI provided by the CUPS server is `ipp://cups-host/printers/printer-name`
Also the code under the `ch.ethz.vppserver` is used throughout the app and provides an IPP client. Am I wrong?

About the idea you have, this could work on some devices, but fail on others. This is a risk that this wouldn't work on some setups I'm afraid. What do you think?

**Comment by debiantriage** [comment]
Many applications send PDF (firefox and chrome, for example) directly to the printer. For your printer avahi-browse shows

pdl=application/vnd.hp-PCL,image/urf,image/jpeg

It will not accept PDF format.

My printer has the same "pdl=..." entry in the TXT record. When it is chosen the display says "This printer isn't available right now". The message seems reasonable to me.

-- 
Brian.

**Comment by chrisapps4u** [comment]
In my opinion these printer shouldn't shown as a cups printer. Because they are indeed a ipp-printer, but not managed by a cups server. And the reason that you get the message "This printer isn't available right now" is not by the pdl-field. You get this, because the rd-field don't reference a cups server. At my printer the rd-field is "ipp/printer". Cups servers never begin with "ipp". The always begin with "printers/...". And when you choose this printer and android call `onStartPrinterStateTracking`, the call of `getPrinter()` don't return a printer. And so the printer isn't available. `getPrinter()` can only return a printer of a cups server.

**Comment by chrisapps4u** [comment]
I make now a code check of CUPS. Since release 1.3b1 CUPS implements DNS-SD: [https://github.com/apple/cups/blob/release-1.3b1/scheduler/dirsvc.c](https://github.com/apple/cups/blob/release-1.3b1/scheduler/dirsvc.c). (See also the [CHANGES.txt](https://github.com/apple/cups/blob/839f33fe98e7001a79960983a90172485721214c/CHANGES.txt#L103) and the issue [#1171](https://github.com/apple/cups/issues/1171).) In function `dnssdBuildTxtRecord` in [line 2202](https://github.com/apple/cups/blob/839f33fe98e7001a79960983a90172485721214c/scheduler/dirsvc.c#L2202) you can see, that with the first implementation the spezial property `printer-type` is returned in the mDNS of a cups server. And in release 1.4b1 CUPS implements an backend for DNS-SD: [https://github.com/apple/cups/blob/release-1.4b1/backend/dnssd.c](https://github.com/apple/cups/blob/release-1.4b1/backend/dnssd.c). In this backend CUPS themselve differ other cups printers with `printer-type`: [line 745](https://github.com/apple/cups/blob/cdad2a2a32bc61c9bccd3b345027b9f7602cf1d7/backend/dnssd.c#L745). And this implementation is since the actual release.

You also can see in [CUPS Implementation of IPP](https://www.cups.org/doc/spec-ipp.html) that `printer-type` is not a normal attribute for ipp. It's a spezial attribute from CUPS.

I implement a query of `printer-type` in `MdnsServices.kt`: see #126
It works fine. No other printers are shown in the list now. I think it is a better solution as ask again the cups server for there printers.

**Comment by BenoitDuffez** [comment]
OK so let me rephrase to be 100% sure I understood what you found out.

* From CUPS 1.4, every printer has a `printer-type` attribute that is non null.
* Filtering the mDNS responses to only the ones containing `printer-type` would hide non-printers on CUPS 1.4+ and possibly all devices (including printers) on CUPS <=1.4

CUPS 1.4.0 was released in August 2009 so I think it's safe to assume that no one would have a printer without `printer-type`.

If my understanding is correct, then your fix is most welcome! Please confirm and I'll merge. Thanks again.

**Comment by chrisapps4u** [comment]
It's more better. mDNS was implemented in CUPS 1.3. With mDNS we can't find printers on CUPS < 1.3. And since this Version (see [dirsvc line 2202](https://github.com/apple/cups/blob/839f33fe98e7001a79960983a90172485721214c/scheduler/dirsvc.c#L2202) also the attribute `printer-type` is implemented.

The attribute `printer-type` is read out of `cupsd_printer_t -> type` (see [dirsvc.c line 2196](https://github.com/apple/cups/blob/839f33fe98e7001a79960983a90172485721214c/scheduler/dirsvc.c#L2196)). `cupsd_printer_t` is defined in [scheduler/printers.h](https://github.com/apple/cups/blob/839f33fe98e7001a79960983a90172485721214c/scheduler/printers.h#L57). In line 57 you can see the definition of `type` as `cups_ptype_t`. `cups_ptype_t` is defined in [cups/cups.h line 57](https://github.com/apple/cups/blob/8cb48f789e8e7977adc09c1d97a1aa69cedccd24/cups/cups.h#L57). This definition exists from the first Version of CUPS i found on github and it is also mentioned in the document [CUPS Implementation of IPP](https://www.cups.org/doc/spec-ipp.html) for the first Version of CUPS.

**Conclusion**
- We can't find any printers on CUPS < 1.3 with mDNS
- From CUPS 1.3, every printer has a `printer-type` attribute.

So **every** printer we can find with mDNS has a `printer-type` attribute.

**Comment by BenoitDuffez** [comment]
Hi @chrisapps4u, can we close this?
